src/java/org/apache/cassandra/index/sai/disk/io/IndexFileUtils.java

```diff
@@ -341,8 +341,18 @@ protected void writeSlow(long bytes, int count) throws IOException
         @Override
         public void writeMostSignificantBytes(long register, int bytes) throws IOException
         {
+            assert buffer != null : "Attempt to use a closed data output";
+            boolean shouldUpdateChecksum = true;
+            if (buffer.remaining() < Long.BYTES)
+            {
+                // Avoid double-counting bytes in the checksum - in this path, super.writeMostSignificantBytes will call
+                // it's super, DataOutputPlus.writeMostSignificantBytes, and that laeds to the writeXXX methods which
+                // are overridden in this class will update the checksum.
```

Comment:
So if the remaining buffer is too small to hold a single long, then `super.writeMostSignificantBytes` updates the checksum but otherwise it does not?  If it calls to writeXXX shouldn't it always update the checksum? This existence of two paths looks very suspicious to me ;)